### PR TITLE
Add basic XCM emulator tests

### PR DESCRIPTION
**Pull request by acatangiu**
This PR adds XCM emulator tests for Relay chains and Asset Hub System parachains.

It adds them under new `integration-tests/emulated/` dir which lives at the root of the repo and doesn't pollute the prod networks code/crates/etc in any way. As you can see from the `diff`, no other files outside `integration-tests` are touched by this PR (except workspace `Cargo.toml` and `Cargo.lock`).

Tests added by this PR are rather old/limited, but I will update them when we bump polkadot-sdk crates deps to `v1.4` or `v1.5` release. This will bring in more refined tests for existing scenarios and many new tests for new scenarios such as complex asset-transfers and bridging - sending XCMs over bridge, and asset transfers between AHs on different relays using bridge.

Also after bumping repo to sdk `v1.4` or `v1.5` release, we can also drop 90% of `integration-tests/emulated/common` and get it from `crates.io`.

Fixes https://github.com/polkadot-fellows/runtimes/issues/103

### CODE REVIEW

File: integration-tests/emulated/common/Cargo.toml
Comment (acatangiu) [inline integration-tests/emulated/common/Cargo.toml:7]: After bumping repo to sdk v1.4 or v1.5 release, we can drop 90% of the code in this module and get it from https://crates.io/crates/emulated-integration-tests-common/1.0.0

**Comment by NachoPal** [comment]
@acatangiu XCM emulator test are going to be on its own repository. It is still WIP: https://github.com/paritytech/polkadot-integration-tests/tree/master

**Comment by acatangiu** [comment]
> @acatangiu We are planing to add XCM emulator to its own repository. It is still WIP: https://github.com/paritytech/polkadot-integration-tests/tree/master

Why would they be on a different repo? I mean I don't see any benefits, while seeing massive drawbacks. It's already hard enough to coordinate over `polkadot-sdk` and `polkadot-fellows`, why would we make it even harder?

**Comment by NachoPal** [comment]
> > @acatangiu We are planing to add XCM emulator to its own repository. It is still WIP: https://github.com/paritytech/polkadot-integration-tests/tree/master
> 
> Why would they be on a different repo? I mean I don't see any benefits, while seeing massive drawbacks. It's already hard enough to coordinate over `polkadot-sdk` and `polkadot-fellows`, why would we make it even harder?

Let's continue the conversation here: https://github.com/polkadot-fellows/runtimes/issues/103#issuecomment-1854039080

File: integration-tests/emulated/assets/asset-hub-kusama/src/tests/hrmp_channels.rs
Comment (xlc) [inline integration-tests/emulated/assets/asset-hub-kusama/src/tests/hrmp_channels.rs:65]: nit: those hardcoded numbers are going to be a maintenance nightmare.
The way my Chopsticks e2e tests doing is round the numbers so small adjustment between versions won't impact the test outcome, and use snapshots testing tool to simply the process of updating them.

**Review by xlc — APPROVED** [review]

File: integration-tests/emulated/assets/asset-hub-kusama/src/tests/hrmp_channels.rs
Comment (acatangiu) [reply integration-tests/emulated/assets/asset-hub-kusama/src/tests/hrmp_channels.rs:65]: all of these `Option<Weight>` params used on events asserts [are using `weight_within_threshold` under the hood](https://github.com/acatangiu/runtimes/blob/77da564b3b9bfe6a88642c8fba3d55dccf6a825f/integration-tests/emulated/common/src/impls.rs#L271), where thresholds [currently allow 33% delta](https://github.com/acatangiu/runtimes/blob/13fe2f99dd10823941a59f5ccda78b19d0980cb4/integration-tests/emulated/common/src/constants.rs#L41).

Everybody (including myself) initially thinks it's checking exact weights because of the very specific numbers used. Maybe rounding them up like you suggested is a good idea to hint to the fact they are actually estimates/thresholds.

**Comment by acatangiu** [comment]
@bkchr @ordian this needs another Rank 3 reviewer because it touches CI files.
(also please re-trigger the failed migration as CI was API rate-limited when trying to set up test env)

**Comment by ordian** [comment]
I recall @rphmeier having issues with how those tests were written affecting async backing changes

File: Cargo.lock
Comment (ordian) [inline Cargo.lock:8735]: why are we bringing all polkadot node side crates for this?

**Comment by acatangiu** [comment]
> I recall @rphmeier having issues with how those tests were written affecting async backing changes

I can't think of any issues with async backing - afaik the emulator doesn't build any blocks per-se - it emulates everything underneath the runtime using some dumb buffers - point being to test the runtime behavior, especially XCM message passing and execution.

These tests currently run on Rococo and Westend too, where I was under the impression async backing is already enabled. Or are you talking about some limitation on some yet unreleased async backing feature?

**Comment by joepetrowski** [comment]
The async backing issues should have all been fixed in:

- https://github.com/paritytech/cumulus/pull/2722
- https://github.com/paritytech/cumulus/pull/2800

File: integration-tests/emulated/assets/asset-hub-kusama/src/tests/teleport.rs
Comment (joepetrowski) [inline integration-tests/emulated/assets/asset-hub-kusama/src/tests/teleport.rs:260]: `teleport_assets` is deprecated. I don't think we should have the three tests that use it. Testing `limited_teleport_assets` is what we care about.

File: integration-tests/emulated/assets/asset-hub-polkadot/src/tests/teleport.rs
Comment (joepetrowski) [inline integration-tests/emulated/assets/asset-hub-polkadot/src/tests/teleport.rs:125]: likewise

**Review by joepetrowski — APPROVED** [review]

File: integration-tests/emulated/assets/asset-hub-kusama/src/tests/teleport.rs
Comment (acatangiu) [reply integration-tests/emulated/assets/asset-hub-kusama/src/tests/teleport.rs]: done

File: integration-tests/emulated/assets/asset-hub-polkadot/src/tests/teleport.rs
Comment (acatangiu) [reply integration-tests/emulated/assets/asset-hub-polkadot/src/tests/teleport.rs]: done

**Comment by acatangiu** [comment]
@ordian @bkchr this is good to go

**Review by ordian — APPROVED** [review]
Didn't looked at the tests closely, but since most of it is going to be removed after bump to 1.5 is merged as you say, don't want to block this PR.

**Comment by acatangiu** [comment]
@joepetrowski bot is not counting your review after latest push, can you try to re-approve?

**Comment by acatangiu** [comment]
@bkchr @ggwpez who has merge button?

**Comment by bkchr** [comment]
@acatangiu please be patient :P https://github.com/polkadot-fellows/runtimes/actions/runs/7410194917/job/20161942771

**Comment by acatangiu** [comment]
> @acatangiu please be patient :P https://github.com/polkadot-fellows/runtimes/actions/runs/7410194917/job/20161942771

ah, nice, I thought bot is ignoring me since I'm not part of fellowship, but I was just being impatient 😛

**Comment by bkchr** [comment]
Yeah, we had a lot of jobs running and github action has probably some upper number of parallel runners per project.

**Pull request merged**
